File: engine/internal/world/seed.go
package world

import (
	"time"

	"github.com/martincollado/kubeislands-engine/internal/proto"
)

var seedNamespaces = []proto.Namespace{
	{ID: "core",          Name: "CORE",          Center: [2]float64{0, 0},     Radius: 8, Hue: "#00FFD1", Ingress: true},
	{ID: "web",           Name: "WEB",           Center: [2]float64{20, 0},    Radius: 8, Hue: "#3EF3FF", Ingress: true},
	{ID: "data",          Name: "DATA",          Center: [2]float64{-18, 6},   Radius: 6, Hue: "#FFB800", Ingress: false},
	{ID: "payments",      Name: "PAYMENTS",      Center: [2]float64{18, 18},   Radius: 8, Hue: "#FF3355", Ingress: false},
	{ID: "observability", Name: "OBSERVABILITY", Center: [2]float64{-4, 20},   Radius: 6, Hue: "#8BE8FF", Ingress: false},
}

var seedDeployments = []proto.Deployment{
	{ID: "core-dep0",          NamespaceID: "core",          Name: "api-gateway",  Replicas: 4},
	{ID: "web-dep0",           NamespaceID: "web",           Name: "frontend",     Replicas: 5},
	{ID: "web-dep1",           NamespaceID: "web",           Name: "cdn-proxy",    Replicas: 3},
	{ID: "data-dep0",          NamespaceID: "data",          Name: "postgres",     Replicas: 3},
	{ID: "payments-dep0",      NamespaceID: "payments",      Name: "payment-svc",  Replicas: 4},
	{ID: "payments-dep1",      NamespaceID: "payments",      Name: "fraud-detect", Replicas: 3},
	{ID: "observability-dep0", NamespaceID: "observability", Name: "prometheus",   Replicas: 2},
}

var seedServices = []proto.Service{
	{ID: "core/api-gateway", Name: "api-gateway", Namespace: "core", Kind: "Service", ClusterIP: "10.0.0.1", Type: "ClusterIP", Ports: []string{"80/TCP", "443/TCP"}},
	{ID: "core/grpc-internal", Name: "grpc-internal", Namespace: "core", Kind: "Service", ClusterIP: "10.0.0.2", Type: "ClusterIP", Ports: []string{"9090/TCP"}},
	{ID: "web/frontend", Name: "frontend", Namespace: "web", Kind: "Service", ClusterIP: "10.0.1.1", Type: "LoadBalancer", Ports: []string{"80/TCP", "443/TCP"}},
	{ID: "web/cdn-proxy", Name: "cdn-proxy", Namespace: "web", Kind: "Service", ClusterIP: "10.0.1.2", Type: "ClusterIP", Ports: []string{"8080/TCP"}},
	{ID: "data/postgres", Name: "postgres", Namespace: "data", Kind: "Service", ClusterIP: "10.0.2.1", Type: "ClusterIP", Ports: []string{"5432/TCP"}},
	{ID: "payments/payment-svc", Name: "payment-svc", Namespace: "payments", Kind: "Service", ClusterIP: "10.0.3.1", Type: "ClusterIP", Ports: []string{"8080/TCP"}},
	{ID: "observability/prometheus", Name: "prometheus", Namespace: "observability", Kind: "Service", ClusterIP: "10.0.4.1", Type: "ClusterIP", Ports: []string{"9090/TCP"}},
}

var seedConfigMaps = []proto.ConfigMap{
	{ID: "core/app-config", Name: "app-config", Namespace: "core", Kind: "ConfigMap", DataKeys: []string{"LOG_LEVEL", "MAX_CONNECTIONS", "TIMEOUT_MS"}},
	{ID: "core/feature-flags", Name: "feature-flags", Namespace: "core", Kind: "ConfigMap", DataKeys: []string{"enable_dark_mode", "enable_beta_api", "maintenance_mode"}},
	{ID: "web/frontend-config", Name: "frontend-config", Namespace: "web", Kind: "ConfigMap", DataKeys: []string{"API_BASE_URL", "CDN_ORIGIN", "GA_TRACKING_ID"}},
	{ID: "web/nginx-conf", Name: "nginx-conf", Namespace: "web", Kind: "ConfigMap", DataKeys: []string{"nginx.conf"}},
	{ID: "data/postgres-init", Name: "postgres-init", Namespace: "data", Kind: "ConfigMap", DataKeys: []string{"init.sql", "schema.sql"}},
	{ID: "observability/prometheus-rules", Name: "prometheus-rules", Namespace: "observability", Kind: "ConfigMap", DataKeys: []string{"alert_rules.yml", "recording_rules.yml"}},
}

var seedBridges = []proto.Bridge{
	{A: "core", B: "web",           Traffic: 0.72, ErrorRate: 0.01},
	{A: "core", B: "data",          Traffic: 0.45, ErrorRate: 0.02},
	{A: "core", B: "payments",      Traffic: 0.60, ErrorRate: 0.03},
	{A: "web",  B: "payments",      Traffic: 0.38, ErrorRate: 0.01},
	{A: "core", B: "observability", Traffic: 0.25, ErrorRate: 0.00},
}

func seedHealth(depIdx, podIdx int) proto.Health {
	val := (depIdx*7 + podIdx*13) % 100
	if val < 85 {
		return proto.HealthReady
	}
	if val < 95 {
		return proto.HealthPending
	}
	return proto.HealthFailed
}

func buildSeedPods() []proto.Pod {
	now := time.Now().UnixMilli()
	var pods []proto.Pod
	for di, dep := range seedDeployments {
		for pi := 0; pi < dep.Replicas; pi++ {
			pods = append(pods, proto.Pod{
				ID:           dep.ID + "-pod" + itoa(pi),
				DeploymentID: dep.ID,
				NamespaceID:  dep.NamespaceID,
				Slot:         pi,
				Health:       seedHealth(di, pi),
				CreatedAt:    now - int64(di*3600+pi*300)*1000,
			})
		}
	}
	return pods
}

func itoa(n int) string {
	if n == 0 {
		return "0"
	}
	digits := []byte{}
	for n > 0 {
		digits = append([]byte{byte('0' + n%10)}, digits...)
		n /= 10
	}
	return string(digits)
}

File: src/data/seed.ts
/**
 * Seed data — all K8s entities for the v1.0 world.
 * Hex codes live here as the explicit exception to the "no literals" rule.
 */
import { color } from '@/theme'

export type Health = 'ready' | 'pending' | 'failed'

export interface Namespace {
  id: string
  name: string
  center: [number, number]  // (x, z) — y is always 0
  radius: 6 | 8
  hue: string               // from DESIGN_SYSTEM §1
  ingress: boolean
}

export interface Deployment {
  id: string
  namespaceId: string
  name: string
  replicas: number
}

export interface Pod {
  id: string
  deploymentId: string
  namespaceId: string
  slot: number
  health: Health
  createdAt: number
}

export interface Bridge {
  a: string
  b: string
  traffic: number     // 0..1
  errorRate: number   // 0..1
}

export interface EventLog {
  id: string
  t: number
  verb: 'ADDED' | 'MODIFIED' | 'DELETED' | 'WARN' | 'ERROR'
  namespace: string
  target?: string
  message: string
}

export interface Service {
  id: string
  name: string
  namespace: string
  kind: string
  clusterIP: string
  type: string
  ports: string[]
}

export interface ConfigMap {
  id: string
  name: string
  namespace: string
  kind: string
  dataKeys: string[]
}

export interface Secret {
  id: string
  name: string
  namespace: string
  kind: string
  type: string
}

export interface Job {
  id: string
  name: string
  namespace: string
  kind: string
  status: string
  completions: number
}

export interface CronJob {
  id: string
  name: string
  namespace: string
  kind: string
  schedule: string
  lastRun: number
}

export interface Node {
  name: string
  ready: boolean
  roles: string[]
}

// ─── Namespaces ───────────────────────────────────────────────────────────────
// Centers aligned to hex-ring layout (RING_GAP=22) slots 0-4
// so dynamic namespaces at slots 5+ never overlap.
export const NAMESPACES: Namespace[] = [
  { id: 'core',         name: 'CORE',          center: [0, 0],          radius: 8, hue: color.cyan,   ingress: true  },
  { id: 'web',          name: 'WEB',           center: [19.1, 11.0],    radius: 8, hue: color.cyan2,  ingress: true  },
  { id: 'data',         name: 'DATA',          center: [0, 22],         radius: 6, hue: color.amber,  ingress: false },
  { id: 'payments',     name: 'PAYMENTS',      center: [-19.1, 11.0],   radius: 8, hue: color.red,    ingress: false },
  { id: 'observability',name: 'OBSERVABILITY', center: [-19.1, -11.0],  radius: 6, hue: color.obs,    ingress: false },
]

// ─── Bridges (connectivity per WORLD_DESIGN seed table) ──────────────────────
export const BRIDGES: Bridge[] = [
  { a: 'core', b: 'web',          traffic: 0.72, errorRate: 0.01 },
  { a: 'core', b: 'data',         traffic: 0.45, errorRate: 0.02 },
  { a: 'core', b: 'payments',     traffic: 0.60, errorRate: 0.03 },
  { a: 'web',  b: 'payments',     traffic: 0.38, errorRate: 0.01 },
  { a: 'core', b: 'observability',traffic: 0.25, errorRate: 0.00 },
]

// ─── Deployments ─────────────────────────────────────────────────────────────
export const DEPLOYMENTS: Deployment[] = [
  { id: 'core-dep0',         namespaceId: 'core',          name: 'api-gateway',    replicas: 4 },
  { id: 'web-dep0',          namespaceId: 'web',           name: 'frontend',       replicas: 5 },
  { id: 'web-dep1',          namespaceId: 'web',           name: 'cdn-proxy',      replicas: 3 },
  { id: 'data-dep0',         namespaceId: 'data',          name: 'postgres',       replicas: 3 },
  { id: 'payments-dep0',     namespaceId: 'payments',      name: 'payment-svc',    replicas: 4 },
  { id: 'payments-dep1',     namespaceId: 'payments',      name: 'fraud-detect',   replicas: 3 },
  { id: 'observability-dep0',namespaceId: 'observability', name: 'prometheus',     replicas: 2 },
]

// ─── Pods (85% ready / 10% pending / 5% failed) ──────────────────────────────
// Deterministic seed — same result every boot so tests are stable.
function seedHealth(depIdx: number, podIdx: number): Health {
  const val = ((depIdx * 7 + podIdx * 13) % 100)
  if (val < 85) return 'ready'
  if (val < 95) return 'pending'
  return 'failed'
}

export const PODS: Pod[] = DEPLOYMENTS.flatMap((dep, di) =>
  Array.from({ length: dep.replicas }, (_, pi) => ({
    id:           `${dep.id}-pod${pi}`,
    deploymentId: dep.id,
    namespaceId:  dep.namespaceId,
    slot:         pi,
    health:       seedHealth(di, pi),
    createdAt:    Date.now() - (di * 3600 + pi * 300) * 1000,
  }))
)

// ─── Seed events ─────────────────────────────────────────────────────────────
export const SEED_EVENTS: EventLog[] = [
  { id: 'e0', t: Date.now() - 60000,  verb: 'ADDED',    namespace: 'core',    message: 'pod api-gateway-pod3 scheduled' },
  { id: 'e1', t: Date.now() - 45000,  verb: 'MODIFIED', namespace: 'web',     message: 'deployment frontend scaled 4 → 5' },
  { id: 'e2', t: Date.now() - 32000,  verb: 'WARN',     namespace: 'payments',message: 'pod fraud-detect-pod1 restart count 3' },
  { id: 'e3', t: Date.now() - 18000,  verb: 'ADDED',    namespace: 'data',    message: 'configmap postgres-config-v2 applied' },
  { id: 'e4', t: Date.now() - 9000,   verb: 'ERROR',    namespace: 'payments',message: 'pod fraud-detect-pod2 CrashLoopBackOff' },
  { id: 'e5', t: Date.now() - 4000,   verb: 'MODIFIED', namespace: 'core',    message: 'hpa api-gateway adjusted replicas' },
]
